onnx/defs/math/defs.cc

```diff
@@ -3337,4 +3337,220 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+    if (compute_uv) {
+      *U_shape->add_dim() = dim;
+      *Vh_shape->add_dim() = dim;
+    }
+
+    *S_shape->add_dim() = dim;
+  }
+
+  // SVD Factorization is only applied to the last two dimensions.
+
+  const auto A_n_rows_dim = A_shape.dim(dim_idx);
+  const auto A_n_cols_dim = A_shape.dim(dim_idx + 1);
+
+  const auto min_dimension = A_n_rows_dim.dim_value() < A_n_cols_dim.dim_value() ? A_n_rows_dim : A_n_cols_dim;
+
+  if (compute_uv) {
+    const auto* full_matrices_attr = ctx.getAttribute("full_matrices");
+    const auto full_matrices = full_matrices_attr == nullptr || full_matrices_attr->i() != 0;
+
+    if (full_matrices) {
+      // Full SVD
+      // U and Vh are both square matrices
+      *U_shape->add_dim() = A_n_rows_dim;
+      *U_shape->add_dim() = A_n_rows_dim;
+
+      *S_shape->add_dim() = min_dimension;
+
+      *Vh_shape->add_dim() = A_n_cols_dim;
+      *Vh_shape->add_dim() = A_n_cols_dim;
+    } else {
+      // Partial SVD
+      // U's number of columns becomes the min dimension
+      // Vh's number of rows becomes the min dimension
+      *U_shape->add_dim() = min_dimension;
+      *U_shape->add_dim() = A_n_rows_dim;
+
+      *S_shape->add_dim() = min_dimension;
+
+      *Vh_shape->add_dim() = min_dimension;
+      *Vh_shape->add_dim() = A_n_cols_dim;
+    }
+  } else {
+    *S_shape->add_dim() = min_dimension;
+  }
+}
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SVD,
+    18,
+    OpSchema()
+        .SetDoc(SVD_doc)
+        .Input(
+            0,
+            "A",
+            "Input tensor to factor. Must have at least 2 dimensions.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            // TODO
+            OpSchema::Unknown)
+        .Output(
+            0,
+            "U",
+            "Left singular vectors as matrix. See operator doc for dimension information.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            // TODO
+            OpSchema::Unknown)
+        .Output(
+            1,
+            "S",
+            "Vector of singular values. See operator doc for dimension information.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            // TODO
+            OpSchema::Unknown)
+        .Output(
+            2,
+            "Vh",
+            "Right singular vectors as matrix. See operator doc for dimension information.",
+            "T",
+            OpSchema::Single,
+            true,
+            1,
+            // TODO
+            OpSchema::Unknown)
+        .TypeConstraint(
+            "T",
+            OpSchema::all_numeric_types_with_bfloat_and_complex(),
+            "Constrain input and output types to all numeric tensor types")
+        .Attr(
+            "full_matrices",
+            "If true (1), compute the full SVD. If false (0), compute the reduced SVD. Defaults to true (1).",
+            AttributeProto::INT,
+            false)
```

Comment:
I don't see a boolean type for AttributeProto. Is it acceptable to use an AttributeProto::INT? We treat 0 as false and everything else as true

Reply:
Yes, that's right, the above definition seems fine.